### PR TITLE
2. Add explicit modes for adding and updating data

### PR DESCRIPTION
**Pull request by rt17603**
This is 2 of 3 in a series of PRs to apply the updates discussed in #591. This overall change will update how adding data is treated in the object store.

This PR includes the following changes:

 - Adds in functionality with additional flags:
     -  `if_exists` to deal with situations where new data being added ***overlaps the same time period as current data***
         - ~~None (default)~~ "default" - concatenate new and previous data if there is no time overlap, raise error if new data overlaps previous time period.
         - "new" - just include new data
         - "replace" - *incorporate data into previous data timeseries - this will be more fully implemented in the 3rd and final PR*
    - `save_current` to deal with question around versioning - ***should current data be saved and a new version created?***
       - None (default) - use current version if possible but create new version if data is being changed (i.e. in most cases where if_exists is "new" or "replace")
       - True - save current data and create a new version
       - False - overwrite current data with new data and don't create a new version
    - `force` - whether to replace data even if this is identical to previous (covers previous overwrite case, could be used for potential corrupted data and useful for tests)
    - `overwrite` flag is still present but **deprecated** - will use `if_exists="new"`
 - **Data stored in the object store can be now be deleted as part of adding new data.** This is dependent on the situation and flags passed
    - See new utils/_versions.py `check_if_need_new_version()` function which includes an outline of the decision tree for this.
    - This creates a very short term back up copy of the current data, writes the new data and then deletes the back up data if the write is successful.
 - openghg/store/_datasource.py is where the primary changes are included - mainly to `add_timed_data` and `save` methods. Also fleshed out the `delete_version` commented out stub which already existed.
     - new `_status` attribute added to store details of the data status between adding and saving the data (e.g. overlapping, if_exists etc.)

Main tests added/updated to check this:
- db_integrity/test_data_updates.py
    - test_database_update_force - test identical data can be added using the `force` keyword
    - test_obs_data_read_data_diff - (was previously an xfail) test that new data with same time points but different data values can be added (`if_exists="new"`, `save_current` not set)
    - test_obs_data_read_data_new_version - testing `if_exists="new"`, `save_current=True` combination - expect new version with just new data
    - test_obs_data_read_data_new_version - testing `if_exists="new"`, `save_current=False` combination - expect same version but with just new data
 - store/test_datasource.py
    - test_versioning - updated for situation when creating new version with new data only
    - test_replace_version - added for replacing previous data with new data

*These are being merged into the 'new_object_store' branch for now as a parallel branch as part of the "Special Release" milestone. This branch will then be merged into 'devel' in one go when ready. This also means the CI isn't being run by default but will be run when the overall PR is created against devel.*

### CODE REVIEW

**Review by gareth-j — CHANGES_REQUESTED** [review]
I think this will add some good functionality but needs to use the functios within the `objectstore` submodule to be handling objects in the store instead of using path handling directly.
I'm also not sure I like the use of the `None` as a value for the `if_exists` and `save_current` values. Could these two keyword arguments be combined? At the moment it feels a bit confusing and could lead to bugs if the logic isn't 100% correct.

**Comment by rt17603** [comment]
> I think this will add some good functionality but needs to use the functios within the `objectstore` submodule to be handling objects in the store instead of using path handling directly. I'm also not sure I like the use of the `None` as a value for the `if_exists` and `save_current` values. Could these two keyword arguments be combined? At the moment it feels a bit confusing and could lead to bugs if the logic isn't 100% correct.

- Yeah figuring out how to find the stored objects was one of the tricker parts. Happy to use `objectstore` submodule as needed for path handling - this was originally based on the save module was finding the stored objects but sounds like there is new functionality which now does this better.
- For use of `None` perhaps we could have a third keyword argument instead if you feel this is unclear? The main thing is to make sure there is a default set of steps followed if no options are specified.
- I had a lot of discussions including with Joe around how to describe the process and splitting out the two ideas into separate keywords was the best option we found. I don't think combining these back together would make this less confusing. I think overall it's just a complex set of options we've having to allow for to satisfy both clarity and practicality.

I would point to the design issue around this where I summarised details of discussion and reasoning for the steps chosen https://github.com/openghg/openghg/issues/591.

File: openghg/store/base/_datasource.py [hunk 1]
```diff
@@ -108,17 +116,23 @@ def add_data(
...
+    def add_timed_data(self, data: Dataset, data_type: str, if_exists: Optional[str] = None) -> None:
         """Add data to this Datasource, splitting along the time axis
 
         Args:
             data: An xarray.Dataset
-            data_type: Name of data_type defined by 
+            data_type: Name of data_type defined by
                 openghg.store.spec.define_data_types()
+            if_exists: What to do if existing data is present.
+                - None - checks new and current data for timeseries overlap
```
[the remark this answers is not in the document]
Comment (rt17603) [reply openghg/store/base/_datasource.py:131]: Sure - what do you think would work as a default for this instead? Perhaps if_exists="fail" or "error"?

File: openghg/util/_versions.py [hunk 2]
```diff
@@ -162,3 +164,46 @@ def show_versions(file: IO = sys.stdout) -> None:
...
+             to False, allow previous version to be overwritten.
+
+    Args:
+        if_exists: How to combine new and current data, if present.
+        save_current: Whether to save current data or replace this.
+    Returns:
+        bool: Whether new version should be created
+    """
+    # Determining whether a new version should be created based on inputs.
+    if if_exists is None and save_current is None:
```
[the remark this answers is not in the document]
Comment (rt17603) [reply openghg/util/_versions.py:198]: I included an answer to the general PR response as well but see also issue #591 for discussion around why these arguments were separated out. I would say they end of both being needed to allow for practicality as well as clarity really.

File: openghg/store/base/_datasource.py [hunk 3]
```diff
@@ -231,11 +261,14 @@ def add_timed_data(self, data: Dataset, data_type: str, overwrite: Optional[bool
...
 
+        self._status["updates"] = True
+        self._status["if_exists"] = if_exists
+
     def delete_all_data(self) -> None:
         """Delete the data associated with this Datasource
 
         Returns:
-            None
+            bool: True is all deleted, False otherwise
```
[the remark this answers is not in the document]
Comment (rt17603) [reply openghg/store/base/_datasource.py:271]: Yeah - I briefly forayed into changing this to a bool but then had to switch it back as it didn't play nicely with the cloud aspects - this doc string is one I missed so can change that back.

File: openghg/store/base/_datasource.py [hunk 4]
```diff
@@ -616,29 +678,80 @@ def save(self, bucket: Optional[str] = None, overwrite: Optional[bool] = False)
...
+                overlapping = self._status["overlapping"]
+            else:
+                if_exists = None
+                overlapping = False
+
             # Store the keys for the new data
             new_keys = {}
 
+            version_key = self.define_version_key(version_str)
+            version_folder = Path(bucket) / version_key
```
[the remark this answers is not in the document]
Comment (rt17603) [reply openghg/store/base/_datasource.py:699]: That's fine - this was largely all based on how this was done previously within the `save` function e.g. https://github.com/openghg/openghg/blob/devel/openghg/store/base/_datasource.py#L614
```
data_key = f"{Datasource._data_root}/uuid/{self._uuid}/{version_str}/{daterange}"
```
So how would you want to update that do you think? Would you want to try and incorporate this into `objectstore` in some way perhaps?

File: openghg/store/base/_datasource.py [hunk 5]
```diff
@@ -616,29 +678,80 @@ def save(self, bucket: Optional[str] = None, overwrite: Optional[bool] = False)
...
+
+            # If version is already present, and we are not creating a new version
+            # we may need to delete the previous data.
+            if version_folder.exists() and not new_version:
```
[the remark this answers is not in the document]
Comment (rt17603) [reply openghg/store/base/_datasource.py:703]: From discussion sounds like this is addressed within the __init__.py

File: openghg/store/base/_datasource.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (rt17603) [reply openghg/store/base/_datasource.py:273]: Seems to have been updated anyway as part of merging from devel

**Comment by rt17603** [comment]
@gareth-j - As discussed last week, would you be able to add the explicit changes requested around the datasource key and the save function? If you recall, I followed the original setup present within the `Datasource.save()` but it sounds like `objectstore` supercedes that now so it would be helpful if you could point me to what you would like me to change this to.

**Comment by gareth-j** [comment]
Yeah sure, I'll try and get that done today.

File: openghg/store/base/_datasource.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (rt17603) [reply openghg/store/base/_datasource.py]: This should now be covered by PR #722

File: openghg/store/base/_datasource.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (rt17603) [reply openghg/store/base/_datasource.py]: This should now be addressed by #722

File: openghg/store/base/_datasource.py [hunk 6]
```diff
@@ -616,29 +678,80 @@ def save(self, bucket: Optional[str] = None, overwrite: Optional[bool] = False)
...
+                data_key = self.define_data_key(label=daterange, version=version_str)
 
                 new_keys[daterange] = data_key
                 data = self._data[daterange]
 
-                filepath = Path(f"{bucket}/{data_key}._data")
-                parent_folder = filepath.parent
-                if not parent_folder.exists():
-                    parent_folder.mkdir(parents=True, exist_ok=True)
+                filepath = version_folder / f"{daterange}._data"
```
[the remark this answers is not in the document]
Comment (rt17603) [reply openghg/store/base/_datasource.py]: Updated within PR #722

File: openghg/store/base/_datasource.py [hunk 7]
```diff
@@ -649,6 +762,16 @@ def save(self, bucket: Optional[str] = None, overwrite: Optional[bool] = False)
                 #     set_object_from_file(bucket=bucket, key=data_key, filename=filepath)
                 # path = f"{bucket_path}/data"
 
+            # If write has been successful, remove any back up data.
+            if delete_version:
+                if version_folder_backup.exists():
+                    logger.warning(f"Deleting previous stored data for this version: {version_str}")
+                    self.delete_version(version_str_backup)
+                    os.rmdir(version_folder_backup)
```
[the remark this answers is not in the document]
Comment (rt17603) [reply openghg/store/base/_datasource.py]: Has been updated through PR #722

File: openghg/store/base/_datasource.py [hunk 8]
```diff
@@ -616,29 +678,80 @@ def save(self, bucket: Optional[str] = None, overwrite: Optional[bool] = False)
...
+            else:
+                delete_version = False
+
+            # Create back up of original data in case writing new data is unsuccessful
+            if delete_version:
+                version_str_backup = self.define_backup_version(version_str)
+                version_key_backup = self.define_version_key(version_str_backup)
+
+                # Create full path to back up folder and move current version folder
+                version_folder_backup = Path(bucket) / version_key_backup
```
[the remark this answers is not in the document]
Comment (rt17603) [reply openghg/store/base/_datasource.py]: Has been updated through PR #722

File: openghg/store/_emissions.py [hunk 9]
```diff
@@ -95,17 +106,42 @@ def read_file(
...
+        if overwrite and if_exists == "default":
+            logger.warning("Overwrite flag is deprecated in preference to `if_exists` (and `save_current`) inputs."
+                           "See documentation for details of these inputs and options.")
+            if_exists = "new"
+
+        # Making sure data can be force overwritten if force keyword is included.
+        if force and if_exists == "default":
+            if_exists = "new"
+
+        new_version = check_if_need_new_version(if_exists, save_current)
```
Comment (gareth-j) [inline openghg/store/_emissions.py:143]:
```suggestion
        new_version = check_if_need_new_version(if_exists=if_exists, save_current=save_current)
```

File: openghg/store/_emissions.py [hunk 10]
```diff
@@ -183,16 +226,30 @@ def transform_data(
...
         from openghg.types import EmissionsDatabases
-        from openghg.util import load_emissions_database_parser
+        from openghg.util import load_emissions_database_parser, check_if_need_new_version
+
+        if overwrite and if_exists == "default":
+            logger.warning("Overwrite flag is deprecated in preference to `if_exists` (and `save_current`) inputs."
+                           "See documentation for details of these inputs and options.")
+            if_exists = "new"
+
+        new_version = check_if_need_new_version(if_exists, save_current)
```
Comment (gareth-j) [inline openghg/store/_emissions.py:252]:
```suggestion
        new_version = check_if_need_new_version(if_exists=if_exists, save_current=save_current)
```

File: openghg/store/_eulerian_model.py [hunk 11]
```diff
@@ -76,12 +94,26 @@ def read_file(
...
+                "Overwrite flag is deprecated in preference to `if_exists` (and `save_current`) inputs."
+                "See documentation for details of these inputs and options."
+            )
+            if_exists = "new"
+
+        # Making sure data can be force overwritten if force keyword is included.
+        if force and if_exists == "default":
+            if_exists = "new"
+
+        new_version = check_if_need_new_version(if_exists, save_current)
```
Comment (gareth-j) [inline openghg/store/_eulerian_model.py:108]:
```suggestion
        new_version = check_if_need_new_version(if_exists=if_exists, save_current=save_current)
```

File: openghg/store/_footprints.py [hunk 12]
```diff
@@ -267,10 +286,30 @@ def read_file(
...
+                "Overwrite flag is deprecated in preference to `if_exists` (and `save_current`) inputs."
+                "See documentation for details of these inputs and options."
+            )
+            if_exists = "new"
+
+        # Making sure data can be force overwritten if force keyword is included.
+        if force and if_exists == "default":
+            if_exists = "new"
+
+        new_version = check_if_need_new_version(if_exists, save_current)
```
Comment (gareth-j) [inline openghg/store/_footprints.py:300]:
```suggestion
        new_version = check_if_need_new_version(if_exists=if_exists, save_current=save_current)
```

File: openghg/store/_obscolumn.py [hunk 13]
```diff
@@ -87,6 +104,19 @@ def read_file(
...
+                "Overwrite flag is deprecated in preference to `if_exists` (and `save_current`) inputs."
+                "See documentation for details of these inputs and options."
+            )
+            if_exists = "new"
+
+        # Making sure data can be force overwritten if force keyword is included.
+        if force and if_exists == "default":
+            if_exists = "new"
+
+        new_version = check_if_need_new_version(if_exists, save_current)
```
Comment (gareth-j) [inline openghg/store/_obscolumn.py:118]:
```suggestion
        new_version = check_if_need_new_version(if_exists=if_exists, save_current=save_current)
```

File: openghg/store/_obssurface.py [hunk 14]
```diff
@@ -198,6 +224,19 @@ def read_file(
...
+                "Overwrite flag is deprecated in preference to `if_exists` (and `save_current`) inputs."
+                "See documentation for details of these inputs and options."
+            )
+            if_exists = "new"
+
+        # Making sure data can be force overwritten if force keyword is included.
+        if force and if_exists == "default":
+            if_exists = "new"
+
+        new_version = check_if_need_new_version(if_exists, save_current)
```
Comment (gareth-j) [inline openghg/store/_obssurface.py:238]:
```suggestion
        new_version = check_if_need_new_version(if_exists=if_exists, save_current=save_current)
```

**Review by gareth-j — CHANGES_REQUESTED** [review]
Just a few changes for passing in keyword arguments, otherwise looks good!

File: openghg/store/_emissions.py (hunk 9 above)
Comment (gareth-j) [reply openghg/store/_emissions.py:143]: Yeah all good, we don't need to be too pedantic about it for functions with < 3 arguments.

**Review by gareth-j — APPROVED** [review]
Looks good!

**Pull request merged**